### DISCUSSION TITLE
Allow to generate a batch of building blocks via developer API

### DISCUSSION THREAD

**Discussion by jnnwnk** [elementor/elementor#25411] [Feature Request]
### Prerequisites

- [X] I have searched for similar feature requests in both open and closed discussions and cannot find an existing request.
- [X] I have verified that the feature is still missing from the latest stable versions of Elementor or Elementor Pro.

### What problem is your feature request going to solve? Please describe..

I have a pre-formatted text (e.g., with title, paragraphs) that I want to insert on a website. The text is available in a meta-format (e.g., HTML, JSON) with information about the formatting. I want to copy this text into the elementor Editor and get a rough basic structure with different widgets and containers based on the formatting information. It is important that, for example, paragraphs are recognized and inserted as individual widgets so that I can subsequently insert images or custom formatting with other widgets between thoseparagraphs.

### Describe the solution you'd like.

Through an interface for developers (programming skills are available) that allows not only the creation of *one* individual widgets but also multiple widgets in a single step, plugins could be developed to handle the corresponding generation. For this to work, it would need to be possible to create multiple widgets/building blocks via the currently existing Widget API, or there would need to be an interface elsewhere that enables the generation of elements in the elementor editor.

### Describe alternatives you've considered.

The Wordpress REST API allows the creation of pages with arbitrary content (not just widgets within the editor view) from external sources. If there's a way to convert elements from the Wordpress editor into widgets/containers in the elementor editor, the generation of the basic structure can be done via the Wordpress REST API. Once the page is initially opened for editing with the elementor editor, a simple conversion occurs, and editing can continue in the Elementor editor (since the primary focus is on the initial creation of pages, backward compatibility is not necessarily required from my perspective).

### Agreement

- [X] I agree that my request may be closed without action if it doesn't meet all the requirements.

**Comment by jnnwnk** [comment]
After taking a closer look at the source code myself, I realized that there are already ways to create widgets via the javascript api (https://developers.elementor.com/js-api/js-api-commands/). It would be great if the documentation could be extended accordingly.

**Comment by webdevs-pro** [comment]
```
var container = elementor.getContainer(widgetId);

$e.run('document/elements/create', {
	model: {
		elType: 'widget',
		widgetType: 'heading',
		settings: {
			title: 'Test',
		}
	},
	container: container,
	options: {
		at: 0
	}
});
```
Try this

**Reply by jnnwnk** [reply]
Yes, I figured that one and some others out. Thanks. 🙂 
But for heading for example I am still struggling on setting the selectable html-tag (it is `h2` by default, but could also be `h1` or `h3` etc.).
Do you know about any documentation on those commands?
This discussion would be completely solved if there is any documentation.

**Reply by m1ga** [reply]
you can use `header_size: "h1"` in the settings to change the headline tag. I've inspected the select field and used its data-setting.

But do you know how to find the current active container id?

**Reply by jnnwnk** [reply]
Thanks. If you don't mind to generating a new container you can create one like
```
var widget= api.run('document/elements/create', {
    model: {
        elType: 'container',
    },
    container: elementor.getPreviewContainer(),
});
var container = widget.view.children._views[Object.keys(container.view.children._views)[0]].getContainer();
```

`elementor` is defined like `$e` after the editor finished loading.

`header_size` was a good hint. There are some files where it appears which might be helpful to figure out some settings for other widgets. 🙂

**Reply by m1ga** [reply]
Thanks. The last line could be `elementor.getContainer(widget.id)`. That works fine but to improve it I would prefer it to be in the current selected container, but I can work with that. Thanks :+1: Now I can convert simple HTML text into Elementor widgts and they are inside a container
